### PR TITLE
Enable -Wconversion in window.c

### PR DESCRIPTION
**Pull request by cjoach**
ref #567

### CODE REVIEW

File: src/nvim/window.c [hunk 1]
```diff
@@ -3074,10 +3076,9 @@ int tabpage_index(tabpage_T *ftp)
...
  */
-static int 
-leave_tabpage (
-    buf_T *new_curbuf,        /* what is going to be the new curbuf,
-                                          NULL if unknown */
+static int leave_tabpage(
+    buf_T *new_curbuf,        // what is going to be the new curbuf,
+                              // NULL if unknown
     int trigger_leave_autocmds
 )
```
Comment (splinterofchaos) [inline src/nvim/window.c]: Generally, stylistic changes make reviewing PRs like this harder, but if you want to make this look like neovim code...
```c
/// Prepare for leaving the current tab page.
/// 
/// When autocommands change "curtab" we don't leave the tab page and return false.
/// @warn When true is returned need to get a new tab page very very soon!
/// @param new_curbuf what is going to be the new curbuf or NULL if unknown
/// @oaram trigger_leave_autocmds If true, trigger leave autocomds.
static int leave_tabpage(buf_T *new_curbuf, int trigger_leave_autocmds) {
  ...
}
```
Been a while since I double checked doxygen syntax, and I didn't check the lines are under 80 chars, but this is basically how it should look.

File: src/nvim/window.c [hunk 2]
```diff
@@ -3653,7 +3652,8 @@ static win_T *win_alloc(win_T *after, int hidden)
     win_append(after, new_wp);
 
   new_wp->w_wincol = 0;
-  new_wp->w_width = Columns;
+  assert(Columns <= INT_MAX);
+  new_wp->w_width = (int)Columns;
```
Comment (splinterofchaos) [inline src/nvim/window.c]: I'm starting to wonder if we shouldn't just have a `long_to_int()` that does this assert automatically.

File: src/nvim/window.c [hunk 3]
```diff
@@ -5373,19 +5379,19 @@ int match_add(win_T *wp, char_u *grp, char_u *pat, int prio, int id, list_T *pos
...
   if (pos_list != NULL)
   {
-    linenr_T	toplnum = 0;
-    linenr_T	botlnum = 0;
-    listitem_T	*li;
-    int		i;
+    linenr_T    toplnum = 0;
+    linenr_T    botlnum = 0;
+    listitem_T  *li;
+    int         i;
```
Comment (splinterofchaos) [inline src/nvim/window.c]: We don't align variables like this. https://neovim.io/develop/style-guide.xml#Horizontal_Whitespace

**Comment by splinterofchaos** [comment]
A lot of `int`s moving to `long`s... why? Firstly, our style guide forbids it (https://neovim.io/develop/style-guide.xml#Integer_Types), secondly, a window's height will probably not get anywhere near `INT_MAX` except in some truly pathological conditions. The number of lines in a file, maybe. If we want to change their types, making them `unsigned` would probably be even better.

File: src/nvim/window.c (hunk 1 above)
Comment (cacplate) [inline src/nvim/window.c]: I was stopping myself for doing too much styling but since you propose it. It's better that way

**Comment by cacplate** [comment]
@splinterofchaos I was hoping to save on castings but I overlooked the no longs convention. I can revert if you want and work on more long->int instead of the other way around. The int->long are moslty local variables that deal with longs so I though that it would be better than type casting everywhere. I will start a parallel work to see if I can revert the tendency.

**Comment by cacplate** [comment]
@splinterofchaos Ok, so the three choices I see are either I convert a bunch of local variables to long, I convert a bunch of options to int or I cast everything. Which one is preferable?

**Comment by splinterofchaos** [comment]
**tl;dr**: Use best judgement.

We seem to have two conflicting aesthetic goals here: Change types when appropriate and avoid casts when possible. If the original type is `long`, then we **must** change it eventually, as per our style guide. If a change causes a superfluous number of casts, then we have to ask whether that change was correct, or whether the types of the variables it interacts with had appropriate types to begin with.
```
// i should be a size_t:
for (int i=0; i < strlen(s); i++) { ... }

void f(linenr_T cols) {
  // i should be a linenr_T.
  for (int i=0; i < cols; i++) { ... }
}

void g(long string_size) {
  // i and string_size should be size_t
  for (long i=0; i < string_size; i++) { ... }
}
```
Changing a type usually means introducing casts and often requires new casts in dependent modules, so we have to accept it as a fact of life. After we sweep through all the code, these casts (hopefully) become garbage that must be collected. IMO, the question isn't really which strategy we should use to do this, but rather where to start, what is the most logical type for each variable, and how to do this safely without introducing new bugs.

**Comment by cacplate** [comment]
Ok so I think I found the middle ground that minimise the casts without converting everything to long. I avoided cascading on everything to limit the size of the PR.

**Comment by cacplate** [comment]
This PR will, I hope, repair #3691 and #3787 so is to be prioritized on them if possible. I made a big change in the handling of numeric options, they now are handled as `int` since it didn't seems that any of them were over `INT_MAX`.

Some options are still of type `long`, it doesn't seem to cause problem but just in case, do some of you think it's a good idea to change them now?

File: src/nvim/ex_getln.c [hunk 4]
```diff
@@ -2635,12 +2635,12 @@ static void cursorcmd(void)
 
   if (cmdmsg_rl) {
     msg_row = cmdline_row  + (ccline.cmdspos / (int)(Columns - 1));
-    msg_col = (int)Columns - (ccline.cmdspos % (int)(Columns - 1)) - 1;
+    msg_col = Columns - (ccline.cmdspos % (int)(Columns - 1)) - 1;
```
Comment (cacplate) [inline src/nvim/ex_getln.c]: removing the int cast

File: src/nvim/option.c [hunk 5]
```diff
@@ -1458,12 +1454,12 @@ do_set (
...
                * use the global value. */
-              if ((long *)varp == &curbuf->b_p_ul
-                  && opt_flags == OPT_LOCAL)
+              if ((int *)varp == &curbuf->b_p_ul && opt_flags == OPT_LOCAL) {
                 value = NO_LOCAL_UNDOLEVEL;
-              else
-                value = *(long *)get_varp_scope(
-                    &(options[opt_idx]), OPT_GLOBAL);
+              } else {
+                value = *(long *)get_varp_scope(&(options[opt_idx]),
```
Comment (cacplate) [inline src/nvim/option.c]: changed to ask int and cast to long

File: src/nvim/window.c [hunk 6]
```diff
@@ -352,7 +351,7 @@ do_window (
     if (Prenum)
       g_do_tagpreview = Prenum;
     else
-      g_do_tagpreview = p_pvh;
+      g_do_tagpreview = (int)p_pvh;
```
Comment (cacplate) [inline src/nvim/window.c]: p_pvh is now `int`. this change is reverted.

**Comment by bfredl** [comment]
Changing all numerical options to `int` (and the range of all options is effectively changed in this PR by the change in `set_num_option`) is by a much bigger change than fixing conversion in `window.c`. Personally I'm not against the change, but I think this should be broken out to a dedicated PR or least have the PR title updated, to attract broader discussion (to see if someone have a legitimate case for an option value bigger than `INT32_MAX`)

**Comment by ZyX-I** [comment]
@bfredl There is `maxmem*` option family which is legible to contain any number which fits into `size_t`. So I think that neither `int` nor `long` is the correct type for numeric options: one should use `ptrdiff_t` (and by no means `ssize_t` here: there are options which may have negative value greater then `-1`: e.g. `&scrolljump` or (GUI) `&guiheadroom`, `&linespace`).

**Comment by ZyX-I** [comment]
Also note that AFAIR code that parses options parses numbers as `long`. This needs to be changed as well (in luaviml I used `intmax_t` and `uintmax_t` in some places AFAIR: better have casts, but only *two* functions for numbers parsing which are guaranteed to satisfy your needs, then have a number of functions with different return value).

**Comment by bfredl** [comment]
good example, even if `maxmem` is in kilobytes ("two terabytes ought to be enough for anyone"). `ptrdiff_t` might have the right semantics but `ptrdiff_t p_someoption // definitely not about pointers` sounds weird, maybe typedef to dedicated num option type?

**Comment by ZyX-I** [comment]
@bfredl 64-bit systems have +32 bits compared to int, and measuring memory in KiB only removes 10 bits. So int is still not enough, same for long because of Windows. Doing `typedef ptrdiff_t NumOpt;`, `typedef int BoolOpt;` (should be changed to `bool` after a while) and `typedef char* StrOpt;` should be a good idea, though I think it is only good if all used option types are `typedef`’ed.

**Comment by cacplate** [comment]
the typedef seems a good idea IMO but it will further enlarge this PR.

@ZyX-I could you ellaborate on
```
Also note that AFAIR code that parses options parses numbers as long. This needs to be changed as well (in luaviml I used intmax_t and uintmax_t in some places AFAIR: better have casts, but only two functions for numbers parsing which are guaranteed to satisfy your needs, then have a number of functions with different return value; not in all places though).
```
I though I got all the code that used `*(long *)`.

**Comment by ZyX-I** [comment]
@cacplate Check code used to parse the options: it is using `vim_str2nr` (`getdigits_int` in some cases, but most of time this is not the case). This is what I mean: you can have `ptrdiff_t` type for integer option, but this is absolutely useless if `vim_str2nr` can only parse to `long`.

**Comment by bfredl** [comment]
I'm working on a PR to refactor bool options to `bool` and it was messier than what I expected. (for instance: one boolean option was actually tristate...) I would really suggest to do the num option refactor in a dedicated PR where both the types and the pointer accesses are changed with no other changes (except changing conversions as needed)

**Comment by cacplate** [comment]
Ok I'll try the `typedef ptrdiff_t NumOpt` in a different PR.
What is the correct flag for this PR for now?

**Comment by cacplate** [comment]
I'm closing this since the majority of the changes depend on the unresolved issue about the numerical option types and I'm not working on it anymore.

**Comment by cacplate** [comment]
I'm closing this since the majority of the changes depend on the unresolved issue about the numerical option types and I'm not working on it anymore.
